MCF Browser: added mcf parsing files

clincoln8 opened the pull request:
graph.js contains Node class and Assertion class to build a local version of the DC Knowledge Graph.
parse-mcf.js contains the ParseMCF class.

pradh commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,216 @@
+        const values = parsePropValues(propValues);
+
+        values.forEach((value) => Assertion.addAssertion(
+            this.curNode, propLabel, value, this.prov));
```
Just to call out:
```
Node: dcid:foo
```
is effectively treated as:
```
Node: dcid:foo
dcid: "foo"
```

pradh commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,216 @@
+        break;
+
+      default:
+        const values = parsePropValues(propValues);
```
The current logic of getNodeFromPropValue() with optional bool, and parsePropValues() feels a bit convoluted to me.

What do think instead of first parsing out the values into either a string or pair of namespace+reference, and then handling Node vs. reference cases separately, instead of combining them in a single function. Something like below (in pseudo code)...

```
  // parsePropValues() returns a list having either a string or <namespace, reference> pair.
  parsedValues = parsePropValues(propValues)
  switch (propLabel) {
    case 'Node':
        // assert len(parsedValues) is 1
        // create new node
        // if value is a pair, ensure namespace = dcid and call setDCID(reference)
    case 'dcid':
        // assert len(parsedValues) is 1
        // assert the value is string
        // do setDCID
     default:
        for (const v of parsedValues) {
           if (v is ns+ref pair) {
             if (ns == LOCAL_NS) {
                // add local reference Assertion
             } else {
                // add DCID reference Assertion
             }
           } else {
              // add value AddAssertion
           }
        }
  }  
```

clincoln8 replied:
sounds good. I am working on this now.

pradh commented on mcf-browser/src/back-end/test/parse-mcf.test.js:
```diff
@@ -0,0 +1,119 @@
+  expect(mcfParser.curNode.dcid).toBe(null);
+});
+
+test('testing parseLine: ', () => {
```
It might be nice to have a e2e test that adds a block of MCF string and checks how Node looks as a result.

pradh reviewed (commented):
Partial review.

I only looked at mcf parsing logic. Yet to look at graph.

I've not done much JS coding, so looping in @beets or @shifucun for JS readability.

clincoln8 commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,216 @@
+export {
+  ParseMCF,
+  shouldReadLine,
+  getPropLabel,
```
The only one that needs to be exported is ParseMCF class. I exported the other functions for testing purposes. Is there a better way to export/import so the test files have access to the functions?

clincoln8 commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,216 @@
+    if (!shouldReadLine(line)) return;
+
+    const propLabel = getPropLabel(line);
+    if (!propLabel) return;
```
I have not implemented anything to handle mcf syntax errors yet. I was thinking maybe we could log the lines that have problems, and then displaying the list of problematic lines/line numbers when finished parsing? 
Or should the parsing stop as soon as we reach the first mcf syntax error?

pradh replied:
Minimally, user should see some indication of failure; better if it's accompanied by a useful error message; even better if we list all problems. Let's start with the easiest first, report useful error on first failure with line number, and then improve it to logging all lines?

clincoln8 replied:
sounds good

beets reviewed (commented):
I only mostly looked at js style and less at the logic. The code is very clean and well documented! Did you consider using typescript? You already have really strong typing information in the documentation.

Btw, we use eslint to format our js files in the website repo. You could try applying it here.

clincoln8 commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,216 @@
+    const propLabel = getPropLabel(line);
+    if (!propLabel) return;
+
+    const propValues = line.replace(propLabel + ':', '').trim();
```
I am not exactly sure what you mean here.  Do you mean like
propLabel = line.split(':')[0]
propValues = line.split(':')[1:].join('').split(',') ?

pradh replied:
Right I just meant:

```
lhs, rhs = line.split(':', 1)
propVals = rhs.split(',')
```

clincoln8 commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -0,0 +1,350 @@
+        await getRemotePropertyValues(curNode.dcid, label, true)
+            .then((valueList) => {
+              if (!valueList) {
+                console.log('ERROR: could not find value for node: ' +
```
Yes, I have not started working on pushing errors to the UI, I have only been logging in the meantime. I will annotate the code with some TODOs to reflect this.

clincoln8 commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -0,0 +1,350 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const API_ROOT = 'https://api.datacommons.org';
```
Thank you for the feedback! I will work on making these changes.

clincoln8 commented:
@pradh I refactored parseMcf to hopefully make it less convoluted. For the parseLine() method, I used the logic use suggested with the parsing values and switch statement, then added helper methods in the switch to make it more readable.  
To avoid the exporting for tests only and to clean up the code more, I moved functions outside of classes to utils.js including calls to data commons and functions used by both parseMcf and parseTmcf.

beets commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -13,126 +13,41 @@
 
 /** Class representation of a single Node in the KG. */
 class Node {
+  localId;
```
it would be helpful to get some descriptions and types of each class variable

beets commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -13,126 +13,41 @@
-  }
-  return value;
-}
+import * as utils from './utils.js'
```
It's best to import only what you need from the class (so it's easy to trace the definitions as one reads the code as well).

beets commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -302,12 +224,16 @@ class Node {
-/** Class representation of a single Assertion (triple) in the KG. */
+/** Class representation of a single Assertion or triple in the KG. */
 class Assertion {
+  src;
```
Ditto comment on class variable documentation

beets commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -13,125 +13,143 @@
 
-const NAMESPACES = ['l', 'schema', 'dcs', 'dcid'];
+import {Assertion, Node} from './graph.js';
+import * as utils from './utils.js'
```
Please import only what you need from utils.js

beets commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -140,77 +158,80 @@ class ParseMCF {
 
-    const propLabel = getPropLabel(line);
-    if (!propLabel) return;
+    const propLabel = line.substring(0, line.indexOf(':')).trim()
```
you can also use str.split(":", 1) here (avoids calling indexof twice)

beets reviewed (commented):
thanks for making the changes! i didn't look too closely at the logic, mostly read through to catch javascript things.

btw, have you tried running eslint?

clincoln8 commented:
@beets Thank you for the feedback! I had tried running eslint before and thought it had run successfully, but clearly not.  I think I have it figured out now. 
Before I was using the default parser which gave an error with the class variable declarations. I did not realize that the parsing error was suppressing other linting errors. I switched to a different one for now which is able to parsed the variable declarations and fix other errors. Is there a specific .eslintrc.js should be using? I am extending to google config in what I am using now.

pradh commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,258 @@
+      }
+    }
+
+    this.curNode = Node.getNode(nodeId, /* shouldCreate */ true);
```
1. In the main parser, for `Node: dcid:foo`, the local-id is `dcid:foo` and DCID is `foo`.
2. We want to treat only `dcid:` prefix in this case (it is a short-form of defining the dcid property, rather than the namespace).
3. Finally, any other namespace prefix (l, dcs, schema) would be invalid.

Considering all of the above, I might suggest something like this:

```
let dcid = '';

if (parsedValues[0] instanceof Object) {
  let ns = parsedValues[0]['ns'];
  if (ns === 'dcid') {
    dcid = parsedValues[0]['ref'];
  } else if (NAMESPACES.includes(ns)) {
    // throw error
  }
}

this.curNode = Node.get(nodeId, true);
if (dcid !== '') {
  this.curNode.setDCID(dcid);
}
```

clincoln8 replied:
when the main parser uses 'dcid:foo' as the local-id for the node, does that mean that other nodes could refer to the node with 'l:dcid:foo' ? Or would other nodes solely use the dcid to link to foo?

pradh commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,258 @@
+   */
+  setCurNodeDCID(parsedValues) {
+    if (!this.curNode || parsedValues.length !== 1 ||
+        typeof parsedValues[0] !== 'string') {
```
Have a TODO to handle each of these case separately and return meaningful errors?

(no repeated values vs. value must be string not ref, etc.)

pradh commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,258 @@
+    for (const val of parsedValues) {
+      let target = val;
+      if (val instanceof Object) {
+        target = Node.getNode(val['ref'], /* shouldCreate */ true);
```
It is just a reference here. So I wonder why we are creating a node?

pradh commented on mcf-browser/src/back-end/utils.js:
```diff
@@ -0,0 +1,115 @@
+ */
+function getValueFromValueObj(valueObj) {
+  if (!('dcid' in valueObj || 'value' in valueObj)) {
+    throw new Error('ERROR remote fetch no dcid or value prop: ' + valueObj);
```
Just to confirm, is "dcid" or "value" property (as in, DC property)? Or are they fields in the object?

clincoln8 replied:
They are fields in the object returned from the DC Rest API. If the target points to another node, then the returned object has a 'dcid' field. Otherwise, if the target is just a text based property/single value, then the returned object has a 'value' field. 
Should I make this more clear in the function docs and/or the error message?

pradh replied:
Got it. I was puzzled by the use of "prop" since it has a meaning wrt DC schema. Perhaps a slight tweak to the error message might help:

  'ERROR: DC API returned an object with no "dcid" or "value" field: ' + valueObj

pradh commented on mcf-browser/src/back-end/utils.js:
```diff
@@ -0,0 +1,115 @@
+ * @return {Promise<boolean>} Returns true if given dcid is in any triples in
+ *     Data Commons Knowledge Graph.
+ */
+async function getExistsInKG(dcid) {
```
nit: doesExistInKG(dcid) reads better?

pradh commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -0,0 +1,318 @@
+   *     does not already exist.
+   * @return {Node|null} The found node if it exists or is created.
+   */
+  static getNode(id, shouldCreate) {
```
Which getNode caller passes shouldCreate=false? (My search doesn't find a non-test caller)

clincoln8 replied:
there had been some front end callers that had called it with false, but I recently changed the implementation so that getNode is never actually called with shouldCreate=false. I'll remove the unused arg, thanks for pointing this out!

pradh commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -0,0 +1,318 @@
+    let localRef = '';
+
+    if (!this.dcid || this.dcid !== this.localId) {
+      localRef = '[l:' + this.localId + ']';
```
nit: do we need [] when dcid is missing?

clincoln8 replied:
not necessarily. I had kept it in just to make it very clear when a dcid was missing. For example in a list of 
(id1, id2 [l:localId], l:localId2)  I thought if we kept the brackets around the last id, then it makes it more obvious that the dcid is missing. 
I will remove the brackets when the dcid is missing if you think that would be better.

pradh commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -0,0 +1,318 @@
+   */
+  setDCID(dcid) {
+    const remote = Node.nodeHash[dcid];
+    if (remote) {
```
Is it more accurate to check for remote != this ? e.g., calling setDCID twice on the same node

clincoln8 replied:
should I throw some sort of error or warning if the user is trying to set a dcid on a node that already has a different dcid? for example. if there is an error in mcf similar to
Node: foo
dcid: remoteFoo
dcid: alsoRemoteFoo

the merge function checks to make sure a node is not being merged with itself, should this check be performed in setDCID before calling the merge as well?

pradh commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -0,0 +1,318 @@
+  /**
+   * Sets the property existsInKG to true if the Node has triples in the DC KG.
+   */
+  async setExistsInKG() {
```
Who calls this?

clincoln8 replied:
This is called from the 'front-end" side. I have set up an in between file called server-api.js which is the only file that the 'front-end' files use for access to back-end functions. This server-api.js file serves as wrapper for all of the back-end functionalities, including calling this function. I have not submitted a PR for that code yet.

pradh commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -0,0 +1,318 @@
+
+            valueList.forEach((valueObj) => {
+              const val = getValueFromValueObj(valueObj);
+              const source = isInverse ? val : this;
```
might be good to assert/check that when isInverse is true that val is of type Node.

pradh commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -0,0 +1,318 @@
+   * Stores remote triples as assertions and inverse Assertions of the calling
+   * Node object. Sets the alreadyFetched property to true if data is fetched.
+   */
+  async fetchRemoteData() {
```
to confirm, the caller is not in this PR?

clincoln8 replied:
yes, this is called from the 'front-end" side. I have not submitted a PR for that code yet.

pradh commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -0,0 +1,318 @@
+      invAssertList.push(invAssert);
+      invAssert = invAssert.invNextAssertion;
+    }
+    return invAssertList;
```
Based on what you have so far, I'm not seeing value in making Assertion a linked-list. I thought the same assertion object gets linked as fwd assertion in the src node, and inv assertion in the target node, but doesn't look like that's the case. Is there a reason I'm missing?

If not, feels like we'll simplify code by making assertions/inv-assertions just lists in Node.

clincoln8 replied:
I think we were intending to use a linked list so that triples could render one at a time as they were remotely fetched, instead of having to wait for all of the triples to be fetched before rendering all of them at once. I ended up not using the approach when implementing the front-end (maybe an optimization that could be made in the future?), so I can switch to using arrays instead. If we end up wanting this rendering, then I can switch back to link lists at that point.

pradh reviewed (commented):
Thanks for making the changes!

Code looks clean. Another round of mostly minor comments.

beets commented:
You can use this [.eslintrc.json](https://github.com/datacommonsorg/website/blob/master/static/.eslintrc.json) but extending google's style is fine as well. Basic eslintrc should cover most things.

clincoln8 commented:
In this update, I am storing 'dcid' and 'l' namespaces in the nodeHash to account for the following example:
Node: ref1
dcid: ref2

Node: ref2
the resulting nodeHash for this mcf would be { 'l:ref1':firstNode, 'dcid:ref2':firstNode, 'l:ref2':secondNode}

The Node constructor expects the namespace either 'l' or 'dcid'. In the case of:
Node: dcid: foo1 
the constructor(via getNode) is called with 'l:dcid:foo1' and then setDCID('foo1') is called. 
The param for setDCID should not have the namespace attached since it will only accept dcids. 

While creating nodes in parse-mcf, I convert the 'dcs' and 'schema' namespaces to 'dcid' to keep things simpler when storing ids. 

There can be the case where a node does not have a local id in which localId field of the Node will be null. This is expected and the front end handles this. An example of this would be 
Node: ref1
prop: dcid:remote
This would result in the creation of a node with localId = null, but then setDCID('remote') will be called, setting the dcid. This node would be the target of the assertion created with the 'prop' label. 

Another example of this is while navigating the browser or using the search bar to see nodes that are not in the local file. These are remote only nodes and do not have localIds.

pradh commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,306 @@
+  }
+
+  /**
+   * Parses a string represetnaing a comma separated list of property values
```
typo: representing

pradh commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,306 @@
+    }
+
+    const propLabel = line.split(':', 1)[0].trim();
+    const propValues = line.substring(line.indexOf(':') + 1);
```
trim?

pradh commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,306 @@
+  parsePropValues(propValues) {
+    const values = [];
+    // split propValues on commas which are not enclosed by double quotes
+    for (const propValue of propValues.split(/,(?=(?:(?:[^"]*"){2})*[^"]*$)/)) {
```
todo: i should understand this

pradh commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,306 @@
+    }
+
+    // handle case: Node: localRef, which means parsedValues[0]==='localRef'
+    let nodeRef = parsedValues[0];
```
nit: feels a bit strange to change the type of nodeRef in some cases. how about doing nodeRef = parsedValues[0] in the else:

```
  if (parsedValues[0] instanceof Object) {
    ...
  } else {
    nodeRef = parsedValues[0];
  }
```

pradh commented on mcf-browser/src/back-end/parse-mcf.js:
```diff
@@ -0,0 +1,306 @@
+    this.curNode = Node.getNode(nodeId);
+
+    if (ns === 'dcid:') {
+      this.curNode.setDCID(nodeRef);
```
assert return value to be true?

pradh commented on mcf-browser/src/back-end/graph.js:
```diff
@@ -0,0 +1,290 @@
+
+  /**
+   * Returns a node with the given ID, creates a new node if shouldCreate is
+   * true, or null if the node does not exist and a node should not be created.
```
comment needs update?  could you point out why shouldCreate isn't needed anymore (i.e., are there no call sites that lookup an existing node)?

pradh reviewed (approved):
Thanks for making all the changes. This is looking great. Approved, with some final nits.
